routing_algorithms/map_matching.hpp

```diff
@@ -0,0 +1,627 @@
+
+        // search from s and t till new_min/(1+epsilon) > length_of_shortest_path
+        while (0 < (forward_heap.Size() + reverse_heap.Size()))
+        {
+            if (0 < forward_heap.Size())
+            {
+                super::RoutingStep(
+                    forward_heap, reverse_heap, &middle_node, &upper_bound, edge_offset, true);
+            }
+            if (0 < reverse_heap.Size())
+            {
+                super::RoutingStep(
+                    reverse_heap, forward_heap, &middle_node, &upper_bound, edge_offset, false);
+            }
+        }
+
+        double distance = std::numeric_limits<double>::max();
+        if (upper_bound != INVALID_EDGE_WEIGHT)
+        {
+            std::vector<NodeID> packed_leg;
+            super::RetrievePackedPathFromHeap(forward_heap, reverse_heap, middle_node, packed_leg);
+            std::vector<PathData> unpacked_path;
+            PhantomNodes nodes;
+            nodes.source_phantom = source_phantom;
+            nodes.target_phantom = target_phantom;
+            super::UnpackPath(packed_leg, nodes, unpacked_path);
+
+            FixedPointCoordinate previous_coordinate = source_phantom.location;
+            FixedPointCoordinate current_coordinate;
+            distance = 0;
+            for (const auto &p : unpacked_path)
+            {
+                current_coordinate = super::facade->GetCoordinateOfNode(p.node);
+                distance += coordinate_calculation::great_circle_distance(previous_coordinate,
+                                                                          current_coordinate);
+                previous_coordinate = current_coordinate;
+            }
+            distance += coordinate_calculation::great_circle_distance(previous_coordinate,
+                                                                      target_phantom.location);
+        }
+
+        return distance;
+    }
+
+    struct HiddenMarkovModel
+    {
+        std::vector<std::vector<double>> viterbi;
+        std::vector<std::vector<std::pair<unsigned, unsigned>>> parents;
+        std::vector<std::vector<float>> path_lengths;
+        std::vector<std::vector<bool>> pruned;
+        std::vector<bool> breakage;
+
+        const Matching::CandidateLists &candidates_list;
+        const EmissionLogProbability& emission_log_probability;
+
+        HiddenMarkovModel(const Matching::CandidateLists &candidates_list, const EmissionLogProbability& emission_log_probability)
+            : breakage(candidates_list.size()), candidates_list(candidates_list), emission_log_probability(emission_log_probability)
+        {
+            for (const auto &l : candidates_list)
+            {
+                viterbi.emplace_back(l.size());
+                parents.emplace_back(l.size());
+                path_lengths.emplace_back(l.size());
+                pruned.emplace_back(l.size());
+            }
+
+            clear(0);
+        }
+
+        void clear(unsigned initial_timestamp)
```

Comment:
- making the parameter a `std::size_t` removes a number of implicit up-casts below.
- would it make sense to use a default parameter of zero?